simulator.py

```diff
@@ -0,0 +1,126 @@
+"""
+Create a simulator to calculate pnl
+"""
+
+import copy
+import uuid
+from thousandaire.dataset import Data
+
+DATASET = 'price_dataset'
+COST_FUNC = 'cost_function'
+
+class Simulator:
+    """
+    Simulate the alpha.
+    """
+    def __init__(self, settings, data, cost_object):
+        self.alpha_formula = None
+        # Means the quantity we have.
+        # Because each instruments may not trade everyday, so we cannot
+        # calculate how many past asset just by yester portfolio.
+        # Thus, we need a argument to save the quantity we have.
+        self.cost_object = cost_object
+        self.data = None
+        self.settings = settings
+        self.__result = Data('result',
+                             ['pnl', 'cost', 'position', 'position_np'])
+        self.__portfolio = list()
+        self.__key = uuid.uuid4()
+        self.decode_data(data)
+
+    def decode_data(self, data):
+        """
+        Decode while in multi-process.
+
+        In multi-process, simulator will get the data reference
+        which alpha need. To prevent polluteing the origin data, we need to
+        copy a local version.
+        """
+        self.data = copy.deepcopy(data)
+
+    def next_date(self):
+        """
+        Move the simulating date.
+        """
+        for dataset in self.data.values():
+            dataset.next_date(self.__key)
+        self.data['target_data'].next_date(self.__key)
+        self.data['workdays'].next_date(auth_key=self.__key)
+
+    def produce_portfolio(self, today):
+        """
+        Find daily portfolio.
+        """
+        portfolio = self.alpha_formula(today, self.data['others'])
+        portfolio.normalize()
+        portfolio.encode_to_nparray(self.settings.target)
+        self.__portfolio.append(portfolio)
```

Comment:
This method is not that large, and is only used by `run`. I would suggest not to make this standalone.